clj/src/bucket_distributor_clj/core.clj

```diff
@@ -0,0 +1,468 @@
+(ns bucket-distributor-clj.core
+  (:require [clojure.tools.logging :as log]
+            [clojure.core.async :as async :refer [chan <!! >!!]]
+            [langohr.core      :as rmq]
+            [langohr.channel   :as lch]
+            [langohr.queue     :as lq]
+            [langohr.exchange  :as le]
+            [langohr.consumers :as lc]
+            [langohr.basic     :as lb]
+            [midje.sweet       :as t])
+  (:import [java.util.concurrent LinkedBlockingQueue TimeUnit Executors ScheduledExecutorService]))
+
+(defn- ex 
+  [msg] (Exception. msg))
+
+(defn require-ch 
+  [conn]
+  (let [ch (lch/open conn)]
+    (when-not ch
+      (throw (Exception. "cannot open channel")))
+    ch))
+
+(defmacro with-conn
+  [binding & body]
+  `(let [config# ~(second binding)
+         ~(first binding) (rmq/connect config#)]
+     (try
+       ~@body
+       (finally 
+         (try
+           (rmq/close ~(first binding))
+           (catch Exception e#
+             (.printStackTrace e#)))))))
+
+(defmacro with-chan
+  [binding & body]
+  `(let [conn# ~(second binding)
+         ~(first binding) (require-ch conn#)]
+     (try
+       ~@body
+       (finally 
+         (try
+           (rmq/close ~(first binding))
+           (catch Exception e#))))))
+
+;;
+;; bucket queue init
+;;
+
+(t/fact 
+  (let [f 1]
+    (+ 2 2) => 4))
+
+(defn- with-rabbit-lock! [conn queue-name instance-id f]
```

Comment:
This is a slow but effective mutual exclusion technique using exclusive queue declaration. This is used to bootstrap the bucket queue in case it doesn't already exist.